### PR TITLE
add `inspect-deal` command to `lotus client`

### PR DESCRIPTION
**Pull request by nonsense**
1. Requires: https://github.com/filecoin-project/go-fil-markets/pull/502
2. Requires: https://github.com/filecoin-project/go-data-transfer/pull/163

---

This PR is adding a `inspect-deal` subcommand to `lotus client`. Its purpose is to provide more information about a specific deal, and filter deals by `deal-id` or by `proposal-cid`.

At the moment the output looks like:

```
Deal ID: 10
Proposal CID: bafyreigvio7johbflmwvcqsqvov26gz62mscy3f5yb4ymjdcbscd2q6d3m

Stage: ReserveClientFunds: Reserving client funds (a few minutes)
  Mar 18 12:18:56 funds reserved, amount <402397026845688>
  Mar 18 12:18:56 adding funds with message <bafy2bzaceax3kesknjfft25kq3m3nnf4zfkuxdebseosessntkbhrmv3sgzgk>
Stage: FundsReserved: FundsReserved (a few minutes)
  Mar 18 12:19:25 initiate data transfer
Stage: StartDataTransfer: Starting data transfer (a few minutes)
  Mar 18 12:19:26 data transfer initiated on channel id <12D3KooWPUTfSmaU5RMKF2oQZfWdDsNGP6ooGoi4wNzKNeGTaDiC-12D3KooWFXYNSdc5z74bjHp1AhVAZR9bijayy62oRRgwnR6fGw3e-0>
              Data transfer stage: Requested
              Data transfer stage: Ongoing
              Mar 18 12:19:26 queued data
              Mar 18 12:19:26 sent data
              Data transfer stage: ResponderCompleted
              Data transfer stage: Completing
Stage: CheckForAcceptance: Checking for deal acceptance (a few minutes)
  Mar 18 12:19:27 provider is verifying the data
  Mar 18 12:19:28 waiting for provider to publish the deal on-chain
  Mar 18 12:21:33 provider has submitted the deal on-chain
  Mar 18 12:22:22 deal has been accepted
Stage: ProposalAccepted: Proposal accepted (a few minutes)
  Mar 18 12:22:22 funds released, amount <402397026845688>
...

```

### CODE REVIEW

File: cli/client.go
Comment (dirkmc) [inline cli/client.go:1121]: Let's add some description here so the user knows they can pass either deal-id or proposal-cid

**Review by dirkmc — APPROVED** [review]
Once the debugging code is removed, LGTM 👍

File: build/params_2k.go
Comment (raulk) [inline build/params_2k.go:43]: Necessary?

File: cli/deal/showdeals.go
Comment (raulk) [inline cli/deal/showdeals.go:1]: Filename should be `inspect.go`.

File: cli/deal/showdeals.go
Comment (raulk) [inline cli/deal/showdeals.go:25]: This could probably be a pointer, and then we can remove `found` and do a nil check below to know if the deal was found or not. Although be careful with local variables in loops.

File: cli/deal/showdeals.go
Comment (raulk) [inline cli/deal/showdeals.go:61]: Ideally we'd do a `fmt.Fprintf` in the `cli.Context.App.Writer`, but if all other commands are writing outputs directly to stdout, then we can keep it like it is.

File: cli/deal/showdeals.go
Comment (raulk) [inline cli/deal/showdeals.go:73]: ubernit: empty line

File: cli/deal/showdeals.go
Comment (raulk) [inline cli/deal/showdeals.go:80]: underscore in var name

File: go.mod
Comment (raulk) [inline go.mod:168]: Remember to update these!

File: node/impl/client/client.go
Comment (raulk) [inline node/impl/client/client.go:219]: Debug code can go.

File: node/impl/client/client.go
Comment (raulk) [inline node/impl/client/client.go:295]: This data can be quite heavy to transfer over the wire (JSON-RPC) if it's not needed.

Since the CLI only allows to inspect a single deal, why not create an `ClientDealInspect` JSON-RPC operation? It's wasteful to send over all the DealStages detail, when we know the CLI will only use a single one.

**Comment by willscott** [comment]
Noting that this is likely in critical path for work next week, so we ideally can get some eyes on it and get it into a version of lotus that a can be run at least on the testnet sooner rather than later.

File: build/params_2k.go
Comment (raulk) [reply build/params_2k.go:43]: This speeds up testing in 2k devnets.

File: node/impl/client/client.go
Comment (raulk) [inline node/impl/client/client.go:200]: debug

**Review by raulk — COMMENTED** [review]
I was trying to take this over the finish line, but I believe it's missing commits. I know we removed the interactive menu, we renamed the command to `inspect-deal`, and it takes `proposal-cid` or `deal-id` flags. But I see none of this in this branch.

File: go.mod
Comment (nonsense) [reply go.mod]: We need to test these first, so no way around not replacing `go-data-transfer` to the one from my branch, as the `go-data-transfer` PR is not merged (same as `go-fil-markets`).

I guess once this is tested we first merge `go-fil-markets`, then `go-data-transfer`, issue releases, update `go.mod` here, and then merge this.

**Comment by nonsense** [comment]
@raulk this is ready for merging, once we get `go-fil-markets` and `go-data-transfer` PRs merged and we update `go.mod` here. The `inspect-deal` has been tested with deals created before/after this version, so it's backward compatible.

**Comment by nonsense** [comment]
I think this is ready for review and merge - all deps have been updated given the latest `go-fil-markets` release.

@dirkmc could you confirm if we should merge this before/after the changes in the `minerx` branch end on `master`?

File: cli/deal/showdeals.go
Comment (nonsense) [reply cli/deal/showdeals.go]: Done.

File: cli/deal/inspect.go
Comment (magik6k) [inline cli/deal/inspect.go:14]: Why a separate function in a separate package? All other commands are just implemented in the cli package.

File: node/impl/client/client.go
Comment (magik6k) [reply node/impl/client/client.go:292]: How big can this get?

File: cli/deal/inspect.go
Comment (nonsense) [reply cli/deal/inspect.go]: Just legacy from the initial UI we were building. I moved to the `cli` package!

File: node/impl/client/client.go
Comment (nonsense) [reply node/impl/client/client.go:292]: If the deal transfer stages are unbounded (i.e. data transfer keeps crashing, and deal is restarted indefinitely), this will be unbounded. In general for successful deals (or those that have a few failures), the struct looks like the following (see attached file).
[output.txt](https://github.com/filecoin-project/lotus/files/6243450/output.txt)

Comment (magik6k) [reply node/impl/client/client.go:292]: So this is ok for a single deal, but I think that we shouldn't set this in `ClientListDeals` (which we do now)

Comment (nonsense) [reply node/impl/client/client.go]: I am now populating the Stages and DealStages fields only when calling `ClientGetDealInfo` and not on `ClientListDeals`.

**Review by magik6k — APPROVED** [review]

**Pull request merged**
